File: Data Structures & Algorithms/same-binary-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {boolean}
     */
    isSameTree(p, q) {
        if (p === q) {
            return true
        }

        if (!p || !q) {
            return false
        }

        if ((p.left && !q.left) || (q.left && !p.left) || (p.right && !q.right) || (q.right && !p.right) ){
            return false
        }
        if (p.left && q.left) {
            if (!this.isSameTree(p.left, q.left)) {
                return false
            }
        }

        if (p.right && q.right) {
            if (!this.isSameTree(p.right, q.right)) {
                return false
            }
        }

        return p.val === q.val 
    }
}

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.cnt = k
        self.res = root.val

        def bst(node):
            if node.left:
                bst(node.left)
            self.cnt -= 1
            if self.cnt == 0:
                self.res = node.val
                return
            if node.right:
                bst(node.right)
                
        bst(root)
        return self.res 